### ISSUE TITLE
Refactor the BlenderBIM Add-on to perform incremental editing

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#1222]
**No new IFC export functionality is planned until this task is complete. Also, no new release will be made until this is fixed.**

This is a tracking issue for the general goal to migrate the BlenderBIM Add-on code to become an incremental editor. It'll take a while until this is closed.

In short, the BlenderBIM Add-on currently functions primarily as an importer and exporter. In other words, it parses the entire IFC dataset atomically, then imports it into Blender. During export, it does the inverse, it parses the entire Blender scene atomically, then exports it into IFC. This creates problems:

 1. Lots of data loss. The Blender data model will never fully match the IFC data model. We do a damn good job already (i.e. visibly better to a user compared to proprietary authoring apps), but we're still missing 30% of the spec.
 2. Long, very inefficient import / export times. So long as this is the case, it becomes a pain to use IFC as a source of truth. If this issue is not solved, it also impacts drawing generation efficiency (see #1153). Drawings are IFC based, and if we cannot very rapidly and continuously export and access the latest IFC data, that slows down our ability to then generate a drawing from it. It also impacts parametric geometry support (half solved, now). Drawings is one example, but the same applies for clash detection, COBie, and everything else.
 3. Lack of code modularity. This creates an unscalable maintenance nightmare, as the entire BIM model is treated atomically.
 4. Very difficult to scale to collaborative editing paradigms - with this new approach, users can only partially load what's interesting, Blender can more easily handle huuuuuuge models, and theoretically work across many apps, and changes even versioned logically without expensive reverse engineered diffs ... this new approach will make it possible in the future to implement multi user editing, not just within Blender, but across other applications too.
 5. Lots of operations cannot be made headless or authoring application agnostic - more and more experience with IfcPatch, IfcSverchok, and seeing what's happening in FreeCAD demonstrate that there is a huge amount of double-up. With an incremental editor, editing is done via ... well, IFC editing! Not translation, so the code can be reused across FreeCAD, NewCloudBIMApp, and so on.
 6. Difficult to write test cases for, as import / export is atomic and depends on Blender. If you can't test it, you can't trust it.
 7. Difficult to onboard new potential developers. Due to the atomic nature of the import / export process, developers will never be able to be sure of the scope and dependency flow of their changes. I've already tried my best to decouple the export code as much as possible, and those familiar with the existing code will realise that I've already decoupled parsing the Blender scene from the actual IFC authoring, so although it looks monolithic and scary, it should be able to be broken down quite easily and also make it easier for future devs.
 8. Errors are only discovered at export time. This can lead to a very slow, ambiguous, and annoying cleanup process, and requires the user to have previous knowledge of what the exporter expects.
 9. It discourages using dedicated UX/UI for BIM aspects that don't translate well to a Blender interface. For example, spreadsheeting, mass data manipulation, resolving change conflicts, paperspace... this new approach will allow many dedicated tools to augment Blender's capabilities, rather than trying to force everything into the Blender paradigm. This will allow other applications to do what they do best, and Blender to do what it does best, and they can both work together.

For the BlenderBIM Add-on to become the best IFC authoring application, it has to only not only operate using the IFC data model, it has to do it live, decoupled from the full Blender scene. This is what is meant by "incremental editing", or partial editing. In short, the IFC dataset, whether it is atomic or a subset of a greater central repository, shall be held in memory (in a file scenario), or if not, accessible via a standardised API. The import process therefore does not need to theoretically translate the entire model into the Blender environment. It can only partially import / visualise / display what's needed. Then, actions in Blender will edit the IFC dataset on the fly. The export process then does no IFC processing whatsoever, and just turns into a disk write operation (in a file scenario - in a central database scenario, there are options too, but let's keep this simple for now). This means faster imports, ultra-fast exports, no data loss, and the Blender editing functions naturally then segment themselves into modules - a module which only edits geometry, a module which only edits materials, a module which only edits documents, and so on.

There are disadvantages to this approach:

 1. It's a lot of work, and a lot of rework. Basically all the current functionality needs to be rebuilt. (Well, theoretically, mostly refactored, but still a lot of work)
 2. Some UI aspects in Blender may need rethinking. E.g. the simple action of duplicating an object now may create complications. Another example is when the user presses undo. The UX needs to make it clear that Blender is _not_ the BIM model ... Blender is simply a client through which it can view and modify portions of the BIM model.

The good news is that the initial attempts to move into this direction are already positive. The new [full native roundtripping feature](https://community.osarch.org/discussion/215/full-native-representation-roundtripping-in-the-blenderbim-add-on) effectively uses this approach already very successfully. With that option enabled, there is zero geometry data loss. Cool, right? What if that was everything? Also, after commit c844076 with that option enabled, you can edit parametric geometry, and export blazing fast.

**Modules which are basically done** (can always be improved, but enough that it works)

 - Aggregates (medium)
 - Attributes (medium)
 - BCF (easy)
 - BIMTester (easy) (thanks @ihabelaghoury !)
 - Clash (easy)
 - Classification (hard)
 - COBie (easy) (thanks @ihabelaghoury !)
 - Contexts (medium)
 - Constraint (medium)
 - Covetool (easy)
 - CSV (easy) (thanks @ihabelaghoury !)
 - Debug (easy)
 - Diff (easy) (thanks @ihabelaghoury !)
 - Document (medium)
 - Drawing (easy, apparently? Note: does not address #1153 at all - only prevents regressions)
 - Geometry (very hard)
 - Georeferencing (medium)
 - IfcSverchok (easy)
 - Material (hard)
 - Model (medium) (known regression for drawing support - will reimplement in future release)
 - Owner History (medium)
 - Patch (easy)
 - Presentation Layers (medium)
 - Project (medium)
 - Pset (hard)
 - PSet Template Definitions (easy)
 - Qto utilities (easy)
 - Root (medium)
 - Search (easy)
 - Spatial (easy)
 - Style (medium)
 - Type (easy)
 - Unit (easy)
 - Void (medium)

**Modules / functionality that we will delay implementation until a future release**

- BIM Server Library (medium) (minimal existing implementation, likely never used by a single user)
- Group (medium) (minimal existing implementation, so user won't miss anything)
- Structural (very hard) (minimal existing implementation, so user won't miss anything) (ping @ihabelaghoury )

Watch this space.

**Comment by PlumpMath** [comment]
Moult cheer up. We're gonna be on this planet. 
From the unknown cave hive to the hacked free space.
yes~
Happy new year~~~ everybody thanks. :))

**Comment by Moult** [comment]
Ping @fbpyr @Krande and @vinnividivicci - let me know if you'd like to help port over the portions you've helped write code for :)

**Comment by Krande** [comment]
Sure! I am quite busy at the moment, but when I have more time I can start porting over the presentation layer stuff :)

**Comment by Moult** [comment]
Note to self: project module should display schema version to user.

**Comment by Moult** [comment]
Ping @vinnividivicci

**Comment by Moult** [comment]
Ping @s-leger - this refactor will heavily impact Archipack integration. Maybe we can chat more on IRC?

**Comment by CyrilWaechter** [comment]
- [ ] When assigning an IfcClass to an object. It leaves object properties in the UI. Need to manually reselect the object to modify its properties.
- [ ] Bug when assigning a material :
```
Python: Traceback (most recent call last):
  File "/home/cyril/.config/blender/2.91/scripts/addons/blenderbim/bim/module/material/operator.py", line 61, in execute
    "material": self.file.by_id(int(obj.BIMObjectMaterialProperties.material)),
ValueError: invalid literal for int() with base 10: ''

location: <unknown location>:-1
```
Step to reproduce : 
* Assign `IfcDuctSegment` to default cube
* Add a material : 
![image](https://user-images.githubusercontent.com/9622862/105821376-0f96db80-5fbb-11eb-8210-ab7e1d139154.png)

**Comment by CyrilWaechter** [comment]
I got another one but I did nothing. I think the mouse have potentially highlighted something or that it occurs when I unexpectedly selected something. I'll edit if I can reproduce : 
```python
WARN (bpy.rna): ../source/blender/python/intern/bpy_rna.c:1487 pyrna_enum_to_py: current value '0' matches no enum in 'BIMRootProperties', '', 'ifc_predefined_type'
WARN (bpy.rna): ../source/blender/python/intern/bpy_rna.c:1487 pyrna_enum_to_py: current value '0' matches no enum in 'BIMRootProperties', '', 'ifc_predefined_type'
WARN (bpy.rna): ../source/blender/python/intern/bpy_rna.c:1487 pyrna_enum_to_py: current value '0' matches no enum in 'BIMRootProperties', '', 'ifc_predefined_type'
Python: Traceback (most recent call last):
  File "/home/cyril/.config/blender/2.91/scripts/addons/blenderbim/bim/module/pset/ui.py", line 138, in poll
    Data.load(props.ifc_definition_id)
  File "/home/cyril/.config/blender/2.91/scripts/addons/blenderbim/bim/module/pset/data.py", line 23, in load
    cls.add_product_psets(product, product_id)
  File "/home/cyril/.config/blender/2.91/scripts/addons/blenderbim/bim/module/pset/data.py", line 48, in add_product_psets
    cls.add_pset(definition.RelatingPropertyDefinition, product_id)
  File "/home/cyril/.config/blender/2.91/scripts/addons/blenderbim/bim/module/pset/data.py", line 57, in add_pset
    "Properties": cls.get_properties_from_template(pset.Name) or []
  File "/home/cyril/.config/blender/2.91/scripts/addons/blenderbim/bim/module/pset/data.py", line 152, in get_properties_from_template
    data_type = str(IfcStore.get_schema().declaration_by_name(prop_template.PrimaryMeasureType or "IfcLabel"))
  File "/usr/lib/python3.9/site-packages/ifcopenshell/ifcopenshell_wrapper.py", line 1939, in declaration_by_name
    return _ifcopenshell_wrapper.schema_definition_declaration_by_name(self, *args)
RuntimeError: Entity with 'IfcAreaDensityMeasure' not found

location: <unknown location>:-1
```

**Comment by Moult** [comment]
The issue occurs because right now we only have pset template definitions for IFC4, but you are likely on an IFC2X3 file... so it tries to match it the best it can, but IfcAreaDensityMeasure didn't exist on IFC2X3, so ... :)

The way to fully fix it is to generate a pset template definition for IFC2X3 but I haven't gotten around to doing that yet. Applied a quick fix for now.

**Comment by Moult** [comment]
Closing bug. Refactor is basically complete - except for drawing generation, which will be solved as part of #1153.
